### PR TITLE
migrate cmake workflows from CircleCI to GHA

### PR DESCRIPTION
**Pull request by pmeier**
cc @seemethere

### CODE REVIEW

**Comment by pmeier** [comment]
There is still something off with the Windows CPU builds. CMake reports it found VS 2019 https://github.com/pytorch/vision/actions/runs/4958983089/jobs/8872517410?pr=7417#step:9:656. However, in the later build it uses 2016 https://github.com/pytorch/vision/actions/runs/4958983089/jobs/8872517410?pr=7417#step:9:674

**Comment by pmeier** [comment]
Ok, this version mismatch is also present on the CircleCI workflows.

**Comment by pmeier** [comment]
OK, even reinstalling CUDA through our own scripts didn't work: https://github.com/pytorch/vision/actions/runs/4991194607/jobs/8937357744

File: .github/workflows/build-cmake.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      script: |
+        set -euo pipefail
+
+        source packaging/windows/internal/vc_install_helper.sh
+
+        if [[ ${{ matrix.gpu-arch-type }} == cuda ]]; then
+          export CU_VERSION=cu$(echo ${{ matrix.gpu-arch-version }} | sed 's/\.//')
+          echo CU_VERSION="${CU_VERSION}"
+          packaging/windows/internal/cuda_install.bat
+        fi
```
Comment (pmeier) [inline .github/workflows/build-cmake.yml:83]: @atalman Ideally, we don't need to reinstall CUDA here. We only do this to get access to your patch from 9baaf701d890317f74ca529052bce840c6aa5238. Could we include these files in the AMI directly?

File: .github/scripts/cmake.sh [hunk 2]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env bash
```
Comment (pmeier) [inline .github/scripts/cmake.sh:1]: This is mostly https://github.com/pytorch/vision/blob/main/packaging/build_cmake.sh with all the slug taken out.

File: .github/scripts/cmake.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+popd
+
+pushd examples/cpp/hello_world
+python trace_model.py
+mkdir -p build
+mv resnet18.pt build
+popd
+
+# This was only needed for the tracing above
+pip uninstall -y torchvision
```
Comment (pmeier) [inline .github/scripts/cmake.sh:62]: Probably not needed, but we are making sure here that we don't need to have a `torchvision` Python install for anything below.

File: examples/cpp/hello_world/CMakeLists.txt [hunk 4]
```diff
@@ -17,4 +17,4 @@ add_executable(hello-world main.cpp)
 # which also adds all the necessary torch dependencies.
 target_compile_features(hello-world PUBLIC cxx_range_for)
 target_link_libraries(hello-world TorchVision::TorchVision)
-set_property(TARGET hello-world PROPERTY CXX_STANDARD 14)
+set_property(TARGET hello-world PROPERTY CXX_STANDARD 17)
```
Comment (pmeier) [inline examples/cpp/hello_world/CMakeLists.txt:20]: Just here to get clean runs. Separate PR in #7608.

File: test/tracing/frcnn/CMakeLists.txt [hunk 5]
```diff
@@ -10,4 +10,4 @@ find_package(Python3 COMPONENTS Development)
 add_executable(test_frcnn_tracing test_frcnn_tracing.cpp)
 target_compile_features(test_frcnn_tracing PUBLIC cxx_range_for)
 target_link_libraries(test_frcnn_tracing ${TORCH_LIBRARIES} TorchVision::TorchVision Python3::Python)
-set_property(TARGET test_frcnn_tracing PROPERTY CXX_STANDARD 14)
+set_property(TARGET test_frcnn_tracing PROPERTY CXX_STANDARD 17)
```
Comment (pmeier) [inline test/tracing/frcnn/CMakeLists.txt:13]: Same as above.

**Comment by pmeier** [comment]
All CMake workflows on GHA are green now: https://github.com/pytorch/vision/actions/runs/5033885907/jobs/9028384793 :tada:

Do we want to have a grade period here, where we run the builds on CircleCI and GHA? IMO, we don't need to, since they were mostly unmaintained and we thoroughly tested the GHA builds in the process of migrating them.

File: .github/workflows/build-cmake.yml (hunk 1 above)
Comment (pmeier) [reply .github/workflows/build-cmake.yml:85]: See pytorch/test-infra#4189

**Review by atalman — APPROVED** [review]

**Review by atalman — APPROVED** [review]

**Pull request merged**